src/typedefs.jl

```diff
@@ -78,7 +100,7 @@ abstract type AbstractCategoricalArray{T, N, R, V, C, U} <: AbstractArray{Union{
 const AbstractCategoricalVector{T, R, V, C, U} = AbstractCategoricalArray{T, 1, R, V, C, U}
 const AbstractCategoricalMatrix{T, R, V, C, U} = AbstractCategoricalArray{T, 2, R, V, C, U}
 
-struct CategoricalArray{T, N, R <: Integer, V, C, U} <: AbstractCategoricalArray{T, N, R, V, C, U}
+mutable struct CategoricalArray{T, N, R <: Integer, V, C, U} <: AbstractCategoricalArray{T, N, R, V, C, U}
```

Comment:
this might have a potential performance impact, but I guess in normal cases the compiler will be able optimize out getting `refs` field here - right?

Reply:
I'm not sure. I changed a couple of loops to store `refs = A.refs` (all others didn't have this problem), but I wonder what happens when the user calls e.g. `getindex` in a loop, as it accesses `A.refs`. Anyway `getindex` already accesses fields inside the pool, which is mutable, so it shouldn't make a big difference. Still it would be nice to know more about this.

Reply:
> but I wonder what happens when the user calls e.g. `getindex` in a loop, as it accesses `A.refs`

This was the use case I was thinking of as recently I was doing similar tests when changing `GroupedDataFrame` which is mutable. Anyway - I think it is OK to leave it as is. I have made a small test - and it seems that in relevant cases the compiler optimizes this out:
```
julia> using BenchmarkTools

julia> struct A{T}
       x::T
       end

julia> mutable struct B{T}
       x::T
       end

julia> f(x, i) = @inbounds x.x[i]
f (generic function with 1 method)

julia> x = rand(10^6);

julia> a = A(x);

julia> b = B(x);

julia> function g(x)
       s = 0.0
       for i in 1:10^6
           s += f(x, i)
       end
       s
       end
g (generic function with 1 method)

julia> @benchmark g($a)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     1.018 ms (0.00% GC)
  median time:      1.028 ms (0.00% GC)
  mean time:        1.039 ms (0.00% GC)
  maximum time:     1.707 ms (0.00% GC)
  --------------
  samples:          4803
  evals/sample:     1

julia> @benchmark g($b)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     1.019 ms (0.00% GC)
  median time:      1.028 ms (0.00% GC)
  mean time:        1.037 ms (0.00% GC)
  maximum time:     1.705 ms (0.00% GC)
  --------------
  samples:          4814
  evals/sample:     1
```

Reply:
Thanks. I was too lazy to run this benchmark myself. :-p